std/algorithm/searching.d

```diff
@@ -4426,6 +4427,429 @@ private void skipAll(alias pred = "a == b", R, Es...)(ref R r, Es es)
+    import std.traits : CommonType;
+
+    static assert(!(substs.length & 1), "The number of substitutions parameters must be even");
+
+    /**
+      Substitute single values with Compile-Time substitution mappings.
+      Complexity: $(BIGO 1) thanks to D's $(D switch) guaranteeing O(1).
+    */
+    auto substitute(Value)(Value value)
+    if (!isInputRange!Value && !is(CommonType!(Value, typeof(substs[0])) == void))
+    {
+        switch (value)
+        {
+            import std.meta : aliasSeqOf;
+            import std.range : iota;
+            foreach (const i; aliasSeqOf!(iota(0, substs.length / 2)))
+                case substs[2 * i]:
+                    return substs[2 * i + 1];
+
+            default: return value;
+        }
+    }
+
+    /// Substitute single range elements with Compile-Time substitution mappings
+    auto substitute(R)(R r)
+    if (isInputRange!R && !is(CommonType!(ElementType!R, typeof(substs[0])) == void))
+    {
+        import std.algorithm.iteration : map;
+        return r.map!(a => substitute(a));
+    }
+
+    /// Substitute subranges with Compile-Time substitution mappings
+    auto substitute(R)(R r)
+    if (isInputRange!R && !is(CommonType!(ElementType!R, ElementType!(typeof(substs[0]))) == void))
+    {
+        // not implemented yet, fallback to runtime variant for now
+        return .substitute(r, substs);
+    }
+}
+
+/// ditto
+auto substitute(alias pred = (a, b) => a == b, R, Substs...)(R r, Substs substs)
+if (isInputRange!R && Substs.length >= 2)
+{
+    import std.algorithm.iteration : map;
+    import std.functional : binaryFun;
+    import std.range.primitives : ElementType;
+    import std.traits : CommonType;
+    import std.meta : aliasSeqOf;
+    import std.range : iota;
+
+    static assert(!(Substs.length & 1), "The number of substitutions parameters must be even");
+
+    // substitute individual elements
+    static if (!is(CommonType!(ElementType!R, Substs) == void))
+    {
+        enum n = Substs.length / 2;
+        auto replaceElement(E)(E e)
+        {
+            foreach (const i; aliasSeqOf!(iota(0, n)))
+                if (binaryFun!pred(e, substs[2 * i]))
+                    return substs[2 * i + 1];
+
+            return e;
+        }
+        return r.map!(a => replaceElement(a));
+    }
+    // substitute subranges
+    else static if (!is(CommonType!(ElementType!R, ElementType!(Substs[0])) == void)  &&
+                        allSatisfy!(isForwardRange, Substs))
+    {
+        import std.range : take;
+
+        enum n = Substs.length / 2;
+
+        auto replaceElement(E)(E e)
+        {
+            const ptrdiff_t hitNr = e[1];
+            switch (hitNr - 1)
+            {
+                // no hit
+                case -1:
+                    return e[0];
+
+                // all replacements
+                foreach (const i; aliasSeqOf!(iota(0, n)))
+                    case i:
+                        return substs[2 * i + 1].take(size_t.max);
```

Comment:
`Take` has to be used the `SubstituteSplitter` to be as generic as possible and then return type needs to match.